tools/oeedger8r/Emitter.ml

```diff
@@ -967,274 +472,861 @@ let warn_size_and_count_params (fd : func_decl) =
-  let funcs = List.append ufuncs tfuncs in
-  List.iter
-    (fun f ->
-      warn_non_portable_types f ;
-      warn_signed_size_or_count_types f ;
-      warn_size_and_count_params f )
-    funcs
-
-(** Includes are emitted in [args.h]. Imported functions have already
-    been brought into function lists. *)
-let gen_t_h (ec : enclave_content) (ep : edger8r_params) =
-  let oe_gen_tfunc_prototypes (tfs : trusted_func list) =
-    if tfs <> [] then
-      List.map (fun f -> sprintf "%s;" (oe_gen_prototype f.tf_fdecl)) tfs
-    else ["/* There were no ecalls. */"]
+(** Generate the Enclave code. *)
+let gen_enclave_code (ec : enclave_content) (ep : edger8r_params) =
+  (* Generate [args.h] which contains [struct]s for ecalls and ocalls *)
+  let oe_gen_args_header (ec : enclave_content) (dir : string) =
+    let oe_gen_marshal_struct (fd : func_decl) (errno : bool) =
+      let gen_member_decl (ptype, decl) =
+        let aty = get_param_atype ptype in
+        let tystr = get_tystr aty in
+        let tystr =
+          if is_foreign_array ptype then
+            sprintf "/* foreign array of type %s */ void*" tystr
+          else tystr
+        in
+        let need_strlen p =
+          (is_str_ptr p || is_wstr_ptr p) && (is_in_ptr p || is_inout_ptr p)
+        in
+        let id = decl.identifier in
+        [ [sprintf "%s %s;" tystr id]
+        ; (if need_strlen ptype then [sprintf "size_t %s_len;" id] else []) ]
+        |> List.flatten
+      in
+      let struct_name = fd.fname ^ "_args_t" in
+      let retval_decl = {identifier= "_retval"; array_dims= []} in
+      let members =
+        [ ["oe_result_t _result;"]
+        ; ( if fd.rtype = Void then []
+          else gen_member_decl (PTVal fd.rtype, retval_decl) )
+        ; (if errno then ["int _ocall_errno;"] else [])
+        ; flatten_map gen_member_decl (List.map conv_array_to_ptr fd.plist) ]
+        |> List.flatten
+      in
+      [ sprintf "typedef struct _%s" struct_name
+      ; "{"
+      ; "    " ^ String.concat "\n    " members
+      ; sprintf "} %s;" struct_name
+      ; "" ]
+    in
+    let oe_gen_user_includes (includes : string list) =
+      if includes <> [] then List.map (sprintf "#include \"%s\"") includes
+      else ["/* There were no user includes. */"]
+    in
+    let oe_gen_user_types (cts : composite_type list) =
+      if cts <> [] then flatten_map emit_composite_type cts
+      else ["/* There were no user defined types. */"; ""]
+    in
+    let oe_gen_ecall_marshal_structs (tfs : trusted_func list) =
+      if tfs <> [] then
+        flatten_map (fun tf -> oe_gen_marshal_struct tf.tf_fdecl false) tfs
+      else ["/* There were no ecalls. */"; ""]
+    in
+    let oe_gen_ocall_marshal_structs (ufs : untrusted_func list) =
+      if ufs <> [] then
+        flatten_map
+          (fun uf -> oe_gen_marshal_struct uf.uf_fdecl uf.uf_propagate_errno)
+          ufs
+      else ["/* There were no ocalls. */"; ""]
+    in
+    let with_errno =
```

Comment:
This is probably repeated elsewhere now that these functions are scoped under the top-level function.